topology/nhlt/nhlt-processor.c

```diff
@@ -0,0 +1,287 @@
+	if (ret < 0)
+		return ret;
+
+	ret = tplg_config_make_add(&nhlt, data_name, SND_CONFIG_TYPE_COMPOUND, top_data_section);
+	if (ret < 0)
+		return ret;
+
+	ret = tplg_config_make_add(&bytes, "bytes", SND_CONFIG_TYPE_STRING, nhlt);
+	if (ret < 0) {
+		snd_config_delete(nhlt);
+		return ret;
+	}
+
+	ret = snd_config_set_string(bytes, src);
+	if (ret < 0) {
+		snd_config_delete(nhlt);
+		snd_config_delete(bytes);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int add_to_manifest(struct tplg_pre_processor *tplg_pp, char *src)
+{
+	const char *data_name = "nhlt";
+	snd_config_t *manifest;
+	snd_config_t *man_inst;
+	snd_config_t *n;
+	snd_config_t *cfg_new;
+	snd_config_t *data;
+	snd_config_iterator_t i, next;
+	const char *id;
+	char *id_name;
+	int ret, id_num = 0;
+
+	/* find manifest section and its named instance */
+	ret = snd_config_search(tplg_pp->output_cfg, "SectionManifest", &manifest);
+	if (ret < 0)
+		return 0;
+
+	snd_config_for_each(i, next, manifest) {
+
+		n = snd_config_iterator_entry(i);
+
+		if (snd_config_get_id(n, &id) < 0)
+			continue;
+
+		if (!strcmp(id, "sof_manifest"))
+			man_inst = n;
+	}
+
+	/* if manifest not found, just exit */
+	if (!man_inst)
+		return 0;
+
+	ret = add_bytes_to_data_section(tplg_pp, src);
+	if (ret < 0)
+		return ret;
+
+	/* add the reference to the bytes into manifest data */
+	ret = snd_config_search(man_inst, "data", &data);
+	if (ret < 0)
+		return ret;
+
+	snd_config_for_each(i, next, data) {
+		const char *name;
+
+		n = snd_config_iterator_entry(i);
+		if (snd_config_get_string(n, &name) < 0)
+			continue;
+
+		/* item already exists */
+		if (!strcmp(name, data_name))
+			return 0;
+		id_num++;
+	}
+
+	/* add new item */
+	id_name = tplg_snprintf("%d", id_num);
+	if (!id_name)
+		return -ENOMEM;
+
+	ret = snd_config_make(&cfg_new, id_name, SND_CONFIG_TYPE_STRING);
+	free(id_name);
+	if (ret < 0)
+		return ret;
+
+	ret = snd_config_set_string(cfg_new, data_name);
+	if (ret < 0)
+		return ret;
+
+	ret = snd_config_add(data, cfg_new);
+	if (ret < 0)
+		snd_config_delete(cfg_new);
+
+	return 0;
+}
+
+static int print_nhlt_as_hex_bytes(struct nhlt *blob, struct endpoint_descriptor **eps, char *dst)
+{
+	uint8_t *top_p = (uint8_t *)blob;
+	struct endpoint_descriptor *ep;
+	uint8_t *ep_p;
+	int i, j;
+
+	for (i = 0; i < sizeof(struct nhlt); i++) {
+		snprintf(dst, 6, "0x%02x,", *top_p);
+		dst += 5;
+		top_p++;
+	}
+
+	for (i = 0; i < blob->endpoint_count; i++) {
+		ep = eps[i];
+		ep_p = (uint8_t *)ep;
+		printf("ep %i length is %u\n", i, ep->length);
+		for (j = 0; j < ep->length; j++) {
+			snprintf(dst, 6, "0x%02x,", *ep_p);
+			dst += 5;
+			ep_p++;
+		}
+	}
+
+	/* remove the last comma... */
+	dst--;
+	*dst = '\0';
+
+	return 0;
+}
+
+/* called at the beginning of topology per-processing */
+int nhlt_init(struct tplg_pre_processor *tplg_pp)
+{
+	if (!tplg_pp)
+		return -EINVAL;
+
+	/* set default values here */
```

Comment:
is this aTODO?

Reply:
no just to say at this point we initialize and set possible default values for internal structs... but can be removed